### ISSUE TITLE
Purpose

### ISSUE THREAD

**Issue by CMCDragonkai** [svanderburg/nijs#1]
Just wondering if the purpose of this project is to provide a way for javascript to interface/RPC to Nix?

What are you using it for at a high level?

Replace npm?

**Comment by wmertens** [comment]
I ran into https://github.com/NixOS/nix/issues/2232 and I wondered how well NodeJS would fare at implementing a Nix interpreter - it seems that this is it?

I saw that it calls out to `nix-build` to build derivations, and that of course uses the `nix` code to evaluate the expression. I wonder how far NiJS would be from replacing nix-build?

And of course, it would be great to build a language extension for vscode with NiJS - especially something like `prettier`.

**Comment by trusktr** [comment]
> it would be great to build a language extension for vscode with NiJS

This would be possible by simply converting NiJS to TypeScript, which already has all the language support in VS Code.

**Comment by wmertens** [comment]
@trusktr no - I meant, use NiJS to interpret Nix code and provide advanced editing functions like scoped names etc (that might be a disaster due to recursive eval ;) ). I did not mean being able to edit NiJS code, it's already javascript.

**Comment by zseri** [comment]
@wmertens I don't think that would be possible with NiJS, but you could take a look at https://github.com/YZITE/nix2js, which tries to do that somewhat, and might be integrated with e.g. NiJS as a backend/handler runtime, to probably achieve that. (recursive lazy eval is possible in JS, although non-trivial to get right, currently I just resort to let all make all attrset elements lazy, and evaluate the rest instantly (e.g. everything outside, and the stuff directly inside of the rhs of a value "assignment", as soon as it gets accessed)).

**Comment by milahu** [comment]
@wmertens 

> I ran into [NixOS/nix#2232](https://github.com/NixOS/nix/issues/2232) and **I wondered how well NodeJS would fare at implementing a Nix interpreter** - it seems that this is it?

see my [nix-eval-js](https://github.com/milahu/nix-eval-js) for a nix interpreter in javascript

goal: prototype an incremental nix evaluator for [rnix-lsp](https://github.com/nix-community/rnix-lsp) intellisense

**Comment by wmertens** [comment]
@milahu cool! What's the state of your project?

**Comment by milahu** [comment]
1% : D
